File: Striver Important Interview Questions/Day 10 - Recursion and Backtracking/permutations.cpp

// https://leetcode.com/problems/permutations/

class Solution {
public:

    void solve(vector<int>&nums, vector<vector<int>>&ans, int n, int si){

        if(si == n){
            ans.push_back(nums);
            return;
        }

        for(int i = si; i < n; i++){
            swap(nums[i], nums[si]);
            solve(nums, ans, n, si+1);
            swap(nums[i], nums[si]);
        }
    }

    vector<vector<int>> permute(vector<int>& nums) {
        vector<vector<int>> ans;
        int n = nums.size();
        if(n == 0) return ans;

        solve(nums, ans, n, 0);
        return ans;
    }
};

File: Neetcode-150-js/Backtracking/permutations.js

// https://leetcode.com/problems/permutations/description/

/**
 * @param {number[]} nums
 * @return {number[][]}
 */
var permute = function(nums) {
  let ans = [];
  let len = nums.length;
  if(len === 0) return ans;

  var solve = (si) => {

    if(si === len){
        ans = [...ans, [...nums]];
    }

    for(let i = si; i < len; i++){
        [nums[i], nums[si]] = [nums[si], nums[i]];
        solve(si+1);
        [nums[i], nums[si]] = [nums[si], nums[i]];
    }
  } 

  solve(0);
  return ans; 
};